### ISSUE TITLE
Figure out a neater, more flexible interface

### ISSUE THREAD

**Issue by bebraw** [bebraw/react-pagify#12]
The current `Paginator` feels too monolithic for my tastes. Given React's [context](https://facebook.github.io/react/docs/context.html) API has stabilized, we might be able to do something way neater. Consider the following:

```javascript
<Paginator.Context
  segments={segmentize({page: 4, pages: 20, beginPages: 3, endPages: 3})}>
  <Paginator.PreviousPage className="previous" />
  <Paginator.Pages className="pages" />
  <Paginator.NextPage className="next" />
</Paginator.Context>
```

The idea is that when using the component, you actually wrap it into a paginator component of your own and deal with layout there.

Any thoughts?

CC @jacktrades, @rowbare, @nadavspi

**Comment by jacktrades** [comment]
I don't know much about the context API yet, but that does look pretty neat.

It might be nice to offer `segmentize` under the `Paginator` namespace, just to reduce the number of imports for users.  I've been going back-and-forth a bit on whether it would be a good idea to provide a few different segmentation algorithms out-of-the-box, or maybe even a sort-of "segmenter plugin API".

**Comment by bebraw** [comment]
Yeah, good point about namespacing. I suppose something like this would look nice:

```javascript
import {Paginator, segmentize} from 'react-pagify';

...
```

Providing a couple of segmentation algorithms out of the box is a nice idea. We can go with a monolithic build for now. The current minified build is 4.5k so it's far from being bloated. If bloat becomes an issue, we'll just split it up to core and segmentation dists. I doubt it comes to that, though.

**Comment by sapegin** [comment]
Looks nice but where should go ellipsis, active and disabled classes and other options?

**Comment by bebraw** [comment]
> Looks nice but where should go ellipsis, active and disabled classes and other options?

* Ellipsis -> `Paginator.Ellipsis`. It can decide whether to render itself based on the context I think.
* Active/disabled classes -> Push to  `Paginator.Context` and look up from there?
* The rest -> Same idea. If it's a general concept, we push it to context. Else to lower level.

Overall this would seem like a win to me.

**Comment by sapegin** [comment]
So this would be possible too:

```html
<Paginator.Ellipsis className="foo"><span>…</span></Paginator.Ellipsis>
```

**Comment by bebraw** [comment]
That would work, yeah. Paginator.Ellipsis deals with the logic while you can define the presentation like that.

**Comment by bebraw** [comment]
It might be a good idea to prepare presets for Bootstrap and such to try out the semantics and get them nailed down.

**Comment by sapegin** [comment]
I think presets would be too complicated but some examples would be very useful.

**Comment by bebraw** [comment]
Yeah. An example or two would work. Bootstrap and Foundation at least.

**Comment by bebraw** [comment]
@sapegin Here's a starting point for Bootstrap. I might be missing something still:

```javascript
<Paginator.Context
  className="pagination"
  onSelect={a => a}
  segments={segmentize({
    page: 4,
    pages: 20,
    beginPages: 3,
    endPages: 3,
    sidePages: 2
  })}>
  <Paginator.PreviousPage className="previous-page" />
  <Paginator.BeginPages />
  <Paginator.Ellipsis><span>…</span></Paginator.Ellipsis>
  <Paginator.Pages />
  <Paginator.Ellipsis><span>…</span></Paginator.Ellipsis>
  <Paginator.EndPages />
  <Paginator.NextPage className="next-page" />
</Paginator.Context>
```

The cool thing is that this allows us to drop direct dependency on `segmentize`.

**Comment by sapegin** [comment]
How it will decide which ellipsis to show?

Now sure we need a default text for buttons:

```html
<Paginator.PreviousPage className="previous-page" >Previous</Paginator.PreviousPage>
```

**Comment by bebraw** [comment]
> How it will decide which ellipsis to show?

Ah, right. It needs more info since there's a positional dependency... So maybe logic needs to go within `Paginator.BeginPages` and `Paginator.EndPages`. We would pass the ellipsis formatting for both using a prop.

```javascript
<Paginator.PreviousPage className="previous-page">Previous</Paginator.PreviousPage>
```

That looks sensible to me.

--

I just had a look at `segmentize`. I think the interface needs to change. Now it can return something like this:

```javascript
[
  [4, 5, 6],
  [8, 9]
]
```

There's not enough information for those context handlers. I think we need to just return an object instead so we get `beginPages`, `centerPages`, and `endPages`. After that it's trivial to perform the checks we need. Btw, we probably want to name `Paginator.Pages` as `Paginator.CenterPages` to be consistent here.

**Comment by jacktrades** [comment]
I think returning an object from `segmentize` is probably the best option.  I implemented a [proof-of-concept of the fixed width pagination algorithm](https://github.com/unbservices/pagination/blob/master/pagination/fixed.py) I mentioned earlier, and ended up needing `segmentize` to return something more along the lines of:

    [[start], [left], page, [right], [end]]

Returning an object would make it easier to later extend the `segmentize` interface to support this in a non-breaking way.

**Comment by bebraw** [comment]
@jacktrades What if I return something like this from `segmentize`?

```javascript
{
  beginPages: [0, 1, 2],
  leftPages: [4, 5],
  centerPage: [6],
  rightPages: [7, 8]
  endPages: [14, 15]
}
```

Of course some of those lists could be empty depending on the case. This would seem like a versatile output format to me and would map well to the context scheme.

**Comment by jacktrades** [comment]
@bebraw That would certainly work for the fixed-width use case (and any others I can think of off the top of my head).

**Comment by bebraw** [comment]
@jacktrades Ok. Let's go with that. I'll adapt `segmentize` to the scheme first and `react-pagify` after that. We can get back to context talks then.

**Comment by bebraw** [comment]
I updated [segmentize](https://github.com/bebraw/segmentize). Instead of `leftPages`, it uses `previousPages`, `centerPage` is just a value directly, `rightPages` is `nextPages`. Please have a look and let me know what you think. The interesting thing is that doing this simplified the implementation a lot.

**Comment by jacktrades** [comment]
LGTM  That should be a pretty flexible schema for writing segmentizers.

**Comment by bebraw** [comment]
Ok. I'll integrate tomorrow.

**Comment by bebraw** [comment]
I gave it a quick go. See https://github.com/bebraw/react-pagify/tree/context . There's still some wiring left to do. The idea is that you will end up with generic `onSelect` that will get wired for you automatically.

The biggest question mark at the moment is what sort of markup to generate. Should each context handler generate just divs? What would be the most convenient for you to style?

**Comment by bebraw** [comment]
@sapegin @jacktrades There's an initial implementation at the `context` branch now. Please have a look and let me know what you think!

I can still add hooks you might need. It would be good to make sure you can style this through Bootstrap etc. though I don't mind releasing it as is. **segmentize** does most of the hard work here after all. I gave it some TLC (1-indexed, clamps given page to [1, pages] range, no `sidePages` by default).

**Comment by jacktrades** [comment]
I took a quick look but I haven't had a chance to dig into React's Context feature.  To be honest, from what I've seen, it seems a bit too magical for my tastes.  Not to say that this isn't an appropriate use-case, just that I probably can't give you great feedback.

**Comment by bebraw** [comment]
No probs. It's a little magical for sure.

I can make things more explicit if needed. This would mean setting `onSelect` per control for example.

**Comment by bebraw** [comment]
Just got an idea. Maybe we should do `<Paginator.Bind field="startPages" />` (operates within `Paginator.Context`) and `Paginator.Ellipsis` (needs some thinking). Smaller, more powerful, less code to write. :)

**Comment by bebraw** [comment]
I did a heavy pass at the API. See [the README](https://github.com/bebraw/react-pagify/tree/context#react-pagify---simple-pagination-for-react). Implementing `Bind` simplified the API a lot and I figured out a neat way to get `Ellipsis` in. It's starting feel like a solid API, but I might be missing a thing or two.

These changes make the library more meta than before but perhaps that's not a bad thing. People can build more specific components on top of this core.

@sapegin Does this look good enough to fit your needs?

**Comment by sapegin** [comment]
Looks good but not sure I understand what `Bind` do ;-)

**Comment by bebraw** [comment]
@sapegin [Here](https://github.com/bebraw/react-pagify/blob/context/src/index.jsx#L26). It takes the segment from the context and renders UI controls based on that. As long as field name matches to the definition and the value of that field is an array, it works.

**Comment by sapegin** [comment]
Maybe `Segment` would be a better name?

**Comment by bebraw** [comment]
Yeah. I'll rename. That's clearer.

**Comment by bebraw** [comment]
@sapegin Done. I don't mind releasing this and work out possible quirks after. Overall the API feels far better than earlier. More flexible.

**Comment by jacktrades** [comment]
The API looks pretty good to me.  I like the idea of building more specific components on top of this, it seems to be flexible enough to implement most of the pagination schemes I've seen.  For something as (deceptively) "simple" as pagination, I imagine most users will want something with a bit less complexity, but this seems like a good core to build those simpler widgets from.

**Comment by bebraw** [comment]
Yeah. I expect we'll end up with `react-pagify-bootstrap`, `react-pagify-pure`, `react-pagify-foundation`, and whatnot based on the core API. :+1:

I'll wait an "ok" from @sapegin and get this out of the door. :smile:

**Comment by jacktrades** [comment]
@bebraw I was curious if there were other forms of pagination I wasn't thinking about, and came across this [gallery of different types of pagination](https://www.smashingmagazine.com/2007/11/pagination-gallery-examples-and-good-practices/).  It might be worth taking a look at.

**Comment by bebraw** [comment]
@jacktrades That's a cool list.

I don't think we're missing any crucial semantics. Dropdowns and such are easy enough to handle outside of the core. :+1:

**Comment by bebraw** [comment]
Merged and published `1.0.0`.
